rfcs/0004-persistent-ledger-builder-controller.md

```diff
@@ -0,0 +1,67 @@
+
+There may be several points between the locked tip and the best tip, and we may need to examine any such point. Most often, though, we'll want the diff between those extrema, and not for intermediate points. To reach intermediate points, we can replay individual operations from the locked tip (perhaps caching that diff). It's possible to create a series of diffs, or their composition, to reach intermediate points, but diffs will consume considerable memory. Since accessing intermediate points should be relatively rare, we'll maintain just a single diff.
+
+It's useful if there are operations on diffs with nice mathematical properties. For example, there could be an addition operation on diffs that results in the composition of those diffs. To "undo" a diff, one could produce the negation of a diff. These ideas suggests an algebraic group, where the carrier set is the set of diffs. The OCaml representation might be a Map from account keys or public keys to Amount.t could be sufficient to implement such a group discipline. An alternative would be a prefix tree with hashes summarizing accounts that have not changed, so that the root is a valid Merkle hash.
+
+These representation choices involve trade-offs that should be considered. The Map representation contains just key and value data, while the prefix tree representation add summary hashes, which could consume more memory. To find account data with the Map, where the account does not appear in that diff, would mean first consulting the diff, which fails, then making an additional database lookup. That could be slower than a lookup in the prefix tree, where summary hashes could be used to guide the database lookup without a failure step.
+
+There's an existing Sync functor that takes a database module, to perform catchup operations. The diffs just described work also with a database, so a SparseDiff functor that takes a database module would be a way to make this setup work. The database module that's passed in is not copyable, but both the Sync and Diff functors will offer a copyable database.
+
+Later, a cache can be introduced onto the Merkle ledger database, allowing us to perform less database queries for commonly looked up information. The details of this, however, is not addressed as part of this particular RFC.
```

Comment:
That's the idea also in #1073 (should be trivial to implement).

Should that be mentioned here?